core/src/apps/zcash/orchard/signer.py

```diff
@@ -0,0 +1,300 @@
+import gc
+from micropython import const
+from typing import TYPE_CHECKING
+
+from trezor import log
+from trezor.crypto.hashlib import blake2b
+from trezor.enums import RequestType, ZcashSignatureType
+from trezor.messages import TxRequest, ZcashAck, ZcashOrchardInput, ZcashOrchardOutput
+from trezor.wire import DataError
+
+from apps.bitcoin.sign_tx import helpers
+from apps.common.paths import HARDENED
+
+from .. import unified_addresses
+from ..hasher import ZcashHasher
+from ..layout import ConfirmOrchardInputsCountOverThreshold
+from .accumulator import MessageAccumulator
+from .crypto import builder, redpallas
+from .crypto.address import Address
+from .crypto.note import Note
+from .debug import watch_gc_async
+from .keychain import OrchardKeychain
+from .random import BundleShieldingRng
+
+if TYPE_CHECKING:
+    from typing import Awaitable, List
+    from apps.common.coininfo import CoinInfo
+    from apps.bitcoin.sign_tx.tx_info import TxInfo
+    from .crypto.keys import FullViewingKey
+    from ..approver import ZcashApprover
+    from .random import ActionShieldingRng
+
+
+OVERWINTERED = const(0x8000_0000)
+FLAGS = const(0b0000_0011)  # spends enbled and output enabled
+MAX_SILENT_ORCHARD_INPUTS = const(8)
+
+
+def skip_if_empty(func):
+    """
+    A function decorated by this will not be evaluated,
+    if the Orchard bundle is impty.
+    """
+
+    async def wrapper(self):
+        if self.actions_count == 0:
+            return
+        else:
+            await func(self)
+
+    return wrapper
+
+
+class OrchardSigner:
+    def __init__(
+        self,
+        tx_info: TxInfo,
+        seed: bytes,
+        approver: ZcashApprover,
+        coin: CoinInfo,
+        tx_req: TxRequest,
+    ) -> None:
+        assert tx_req.serialized is not None  # typing
+
+        self.inputs_count = tx_info.tx.orchard_inputs_count
+        self.outputs_count = tx_info.tx.orchard_outputs_count
+
+        if self.inputs_count + self.outputs_count > 0:
+            self.actions_count = max(
+                2,  # minimal required amount of actions
+                self.inputs_count,
+                self.outputs_count,
+            )
+        else:
+            self.actions_count = 0
+
+        if self.actions_count == 0:
+            return  # no need to initialize other attributes
+
+        self.tx_info = tx_info
+        self.keychain = OrchardKeychain.from_seed_and_coin(seed, coin)
+        self.approver = approver
+        self.coin = coin
+        self.tx_req = tx_req
+        assert isinstance(tx_info.sig_hasher, ZcashHasher)
+        self.sig_hasher: ZcashHasher = tx_info.sig_hasher
+
+        account = tx_info.tx.account
+        assert account is not None  # typing
+        key_path = [
+            32 | HARDENED,  # ZIP-32 constant
+            coin.slip44 | HARDENED,  # purpose
+            account | HARDENED,  # account
+        ]
```

Comment:
ZIP-32 derivation path is
```
m/32'/coin'/account'
```
It felt more simple to fix the first digit, derive the `coin` digit from the `coin_name` and ask only the third (`account`) digit.

Notice that in case of shielded transaction, account digit also determines the `WalletPathChecker`'s attribute:
https://github.com/trezor/trezor-firmware/blob/583805a6bd42bad112009e333d0b3bb70cef02c3/core/src/apps/zcash/signer.py#L66-L70

Reply:
f1a90fcbc9a32913e8bee4d243bb07b8f74805ec

Reply:
https://github.com/trezor/trezor-firmware/blob/f1a90fcbc9a32913e8bee4d243bb07b8f74805ec/core/src/apps/zcash/orchard/signer.py#L92